Add FormIntrinsicSlot property

435THz opened the pull request:
- Add a helper property called `FormIntrinsicSlot`
  - This property starts not loaded. Its value is calculated as soon as it is read for the first time.
  - This was necessary because a `CharData`'s constructor does not initialize its BaseForm.
- `FormIntrinsicSlot` stores a character's "natural" `BaseIntrinsic[0]` index in respect to its form's intrinsic slots. If a character is promoted, it will always try to assume the intrinsic corresponding to that value, if possible.
  - If the first time a character's `FormIntrinsicSlot` is read is when its `BaseIntrinsic[0]` is valid for its form, its slot number will be read and stored normally
  - If the first time a character's `FormIntrinsicSlot` is read is when its `BaseIntrinsic[0]` is not valid for its form, `FormIntrinsicSlot` will be set to a randomly generted number between 0 and 2 (both included). The invalid intrinsic will not be touched at all, so that characters with invalid intrinsics can still be spawned just fine
- Once `FormIntrinsicSlot` is set, using an Ability Capsule is the only way to change its value during normal gameplay.

435THz commented:
Aims to solve https://github.com/audinowho/PMDODump/issues/478

audinowho commented:
Lazy-loading of FormIntrinsicSlot should be avoided if possible.  What can be done instead is for it to be an ordinary member, and assign -1 on initiation.  Then, in all parts of code that assign the BaseIntrinsic (except for Promote), the FormIntrinsicSlot will need to be backwards-computed.  Places to check will be LearnIntrinsic, as well as all spawning code that assigns BaseIntrinsic directly.  Since dealing with the direct spawns may be messy, feel free to create a new method that encapsulates direct Intrinsic assignment plus its lookup to an index.

You could even go all the way and enforce that any change of the BaseIntrinsic must go through a method that ONLY takes the slot and will forward-calculate what intrinsic that slot leads to, thereby preventing any inconsistency whatsoever (but leaving the existing LearnIntrinsic method there SOLELY for use by the debug operations.  However this has more extensive domino effects on the code so be aware of the work involved.  Even if you do take this on, make it a separate PR from this one.

audinowho commented:
Lazy-loading of FormIntrinsicSlot should be avoided if possible. What can be done instead is for it to be an ordinary member, and assign -1 on initiation. Then, in all parts of code that assign the BaseIntrinsic (except for Promote), the FormIntrinsicSlot will need to be backwards-computed. Places to check will be LearnIntrinsic, as well as all spawning code that assigns BaseIntrinsic directly. Since dealing with the direct spawns may be messy, feel free to create a new method that encapsulates direct Intrinsic assignment plus its lookup to an index.

435THz commented:
Changes applied:
1. Assigned FormIntrinsicSlot to -1 on init
2. Added a method that assigns BaseIntrinsics[0] and FormIntrinsicSlot at the same time
3. Changed any external BaseIntrinsics[0] assignment so that it uses this function.
4. Made Promote calculate BaseIntrinsicSlot if and only if it is still -1 when called.
